itests/direct_data_onboard_test.go

```diff
@@ -102,6 +125,458 @@ func TestOnboardRawPiece(t *testing.T) {
+
+	allocations, err = client.StateGetAllocations(ctx, verifiedClientAddr, types.EmptyTSK)
+	require.NoError(t, err)
+	require.Len(t, allocations, 0)
+	eventsFromMessages := buildActorEventsFromMessages(t, ctx, miner.FullNode)
+	writeEventsToFile(t, ctx, miner.FullNode, eventsFromMessages)
+
+	/* --- Tests for the Actor events API --- */
+	pstring := func(s string) *string { return &s }
+	// Match events from Get API and receipts
+	allEvtsFromGetAPI, err := miner.FullNode.GetActorEvents(ctx, &types.ActorEventFilter{
+		FromBlock: pstring("earliest"),
+		ToBlock:   pstring("latest"),
+	})
+	require.NoError(t, err)
+	matchEvents(t, eventsFromMessages, getEventsArray(allEvtsFromGetAPI))
+
+	// match Miner Actor events from subscription channel and Miner Actor events obtained from receipts
+	var subMinerEvts []types.ActorEvent
+	for evt := range minerEvtsChan {
+		subMinerEvts = append(subMinerEvts, *evt)
+		if len(subMinerEvts) == 4 {
+			break
+		}
+	}
+	var allMinerEvts []types.ActorEvent
+	for _, evt := range eventsFromMessages {
+		if evt.EmitterAddr == miner.ActorAddr {
+			allMinerEvts = append(allMinerEvts, evt)
+		}
+	}
+	matchEvents(t, allMinerEvts, subMinerEvts)
+
+	// Match pre-filled events from sector activated channel and events obtained from receipts
+	var prefillSectorActivatedEvts []types.ActorEvent
+	for evt := range sectorActivatedEvtsCh {
+		prefillSectorActivatedEvts = append(prefillSectorActivatedEvts, *evt)
+		if len(prefillSectorActivatedEvts) == 2 {
+			break
+		}
+	}
+	require.Len(t, prefillSectorActivatedEvts, 2)
+	var sectorActivatedEvts []types.ActorEvent
+	for _, evt := range eventsFromMessages {
+		for _, entry := range evt.Entries {
+			if entry.Key == "$type" && bytes.Equal(entry.Value, sectorActivatedCbor) {
+				sectorActivatedEvts = append(sectorActivatedEvts, evt)
+				break
+			}
+		}
+	}
+	matchEvents(t, sectorActivatedEvts, prefillSectorActivatedEvts)
+
+	// Match pre-filled events from subscription channel and events obtained from receipts
+	allEvtsCh, err := miner.FullNode.SubscribeActorEvents(ctx, &types.SubActorEventFilter{
+		Filter: types.ActorEventFilter{
+			FromBlock: pstring("earliest"),
+			ToBlock:   pstring("latest"),
+		},
+		Prefill: true,
+	})
+	require.NoError(t, err)
+	var prefillEvts []types.ActorEvent
+	for evt := range allEvtsCh {
+		prefillEvts = append(prefillEvts, *evt)
+		if len(prefillEvts) == len(eventsFromMessages) {
+			break
+		}
+	}
+	matchEvents(t, eventsFromMessages, prefillEvts)
+}
+
+func getEventsArray(ptr []*types.ActorEvent) []types.ActorEvent {
+	var evts []types.ActorEvent
+	for _, evt := range ptr {
+		evts = append(evts, *evt)
+	}
+	return evts
+}
+
+func matchEvents(t *testing.T, exp []types.ActorEvent, actual []types.ActorEvent) {
+	// height and tipset cid can mismatch because expected events are sourced using APIs that can put in different tipsets
+	for i := range exp {
+		exp[i].Height = 0
```

Comment:
@rvagg The tipset/height we put in the `buildActorEventsFromMessages ` function below on these events is always one ahead of the actual tipset/height these events are actually emitted at (as recorded by the `CollectEvents` function in lotus/chain/events/index.go).

I tried digging into it and it looks like this is because the ` node.StateSearchMsg` function can return messages from a previous tipset as well. Let me know what you think here. I think this is a non-issue.

Reply:
Oh, that's very interesting. Probably a non issue as far as the tests are concerned but this is really helpful to know as I write docs.